lib/certificate.js

```diff
@@ -0,0 +1,114 @@
+const selfsigned = require('selfsigned');
+const fs = require('fs');
+const { blue } = require('chalk');
+const { promisify } = require('util');
+
+const { getPathFromCwd } = require('../lib/cwd');
+const { hosts } = require('../context');
+const { performance } = require('perf_hooks');
+const track = require('../telemetry');
+
+const exists = promisify(fs.exists);
+const mkdir = promisify(fs.mkdir);
+const unlink = promisify(fs.unlink);
+const writeFile = promisify(fs.writeFile);
+const stat = promisify(fs.stat);
+const readFile = promisify(fs.readFile);
+
+const certificateTtl = 1000 * 60 * 60 * 24;
+
+const createSelfSignedCertificate = () => {
+  console.log(blue`Generating self-signed certificate`);
+  const attributes = [
+    {
+      name: 'commonName',
+      value: hosts[0],
+    },
+  ];
+
+  const altNames = hosts.map((host) => ({
+    type: 2, // DNS
+    value: host,
+  }));
+
+  // V4
+  altNames.push({
+    type: 7, // IP
+    ip: '127.0.0.1',
+  });
+
+  // V6
+  altNames.push({
+    type: 7, // IP
+    ip: 'fe80::1',
+  });
+
+  return selfsigned.generate(attributes, {
+    algorithm: 'sha256',
+    days: 30,
+    keySize: 2048,
+    extensions: [
+      {
+        name: 'basicConstraints',
+        cA: true,
+      },
+      {
+        name: 'keyUsage',
+        keyCertSign: true,
+        digitalSignature: true,
+        nonRepudiation: true,
+        keyEncipherment: true,
+        dataEncipherment: true,
+      },
+      {
+        name: 'subjectAltName',
+        altNames,
+      },
+    ],
+  });
+};
+
+const generateCertificate = async (certificatePath, certificateDirPath) => {
+  const startTime = performance.now();
+
+  const pems = await createSelfSignedCertificate();
+
+  const certificateDirExists = await exists(certificateDirPath);
```

Comment:
You should use the sync version of exists as otherwise the file may/may not exist by the next tick. https://nodejs.org/api/fs.html#fs_fs_exists_path_callback